File: src/providers/php/prestart.pl
#!/usr/bin/env perl

undef $/;

use strict;
use warnings;

use FindBin;
use lib ("$FindBin::RealBin");

use File::Find;
use Nixpacks::Config::Template qw(compile_template);
use Nixpacks::Util::Logger;
use Nixpacks::Util::ChmodRecursive qw(chmod_recursive);
use Nixpacks::Util::Laravel qw(is_laravel check_possible_env_errors);

my $server_logger = Nixpacks::Util::Logger->new("server");

Nixpacks::Util::ChmodRecursive::chmod_recursive("/app/storage") if -e "/app/storage";

if ($#ARGV != 1) {
    print STDERR "Usage: $0 <config-file> <output-file>\n";
    exit 1;
}

if(Nixpacks::Util::Laravel::is_laravel()) {
    Nixpacks::Util::Laravel::check_possible_env_errors("/app");
}

Nixpacks::Config::Template::compile_template($ARGV[0], $ARGV[1]);
my $port = $ENV{"PORT"};
$server_logger->info("Server starting on port $port");

File: src/providers/php/scripts/prestart.mjs
#!/usr/bin/env node
import { compileTemplate } from "./config/template.mjs";
import { e } from "./util/cmd.mjs";
import { checkEnvErrors, isLaravel } from "./util/laravel.mjs";
import Logger from "./util/logger.mjs";
import { access, constants } from 'node:fs/promises'

const prestartLogger = new Logger('prestart');
const serverLogger = new Logger('server');

if (process.argv.length != 4) {
    prestartLogger.error(`Usage: ${process.argv[1]} <config-file> <output-file>`)
    process.exit(1);
}

await Promise.all([
    isLaravel() ? checkEnvErrors('/app') : Promise.resolve(),
    access('/app/storage', constants.R_OK)
        .then(() => e('chmod -R ugo+rw /app/storage'))
        .catch(() => {}),
    compileTemplate(process.argv[2], process.argv[3])
]).catch(err => prestartLogger.error(err));

serverLogger.info(`Server starting on port ${process.env.PORT}`)
